tools/testharness.py

```diff
@@ -1,547 +1,535 @@
-            files = [line[:-1] for line in f.readlines()]
-          except IOError as e:
-            sys.stderr.write("Unable to read tests from file %s: %s" % (from_file, e))
-            sys.exit(1)
-          f.close()
-        else:
-          files = None
-
-        if files:
-          for (subdir, xml_file) in [os.path.split(x) for x in xml_files]:
-            temp_files=files
-            for file in temp_files:
-              if xml_file == file:
-                p = etree.parse(os.path.join(subdir,xml_file))
-                prob_defn = p.findall("problem_definition")[0]
-                prob_nprocs = int(prob_defn.attrib["nprocs"])                
-                testprob = regressiontest.TestProblem(filename=os.path.join(subdir, xml_file),
-                                                      verbose=self.verbose, replace=self.modify_command_line(prob_nprocs), genpbs=genpbs)
-                self.tests.append((subdir, testprob))
-                files.remove(xml_file)
-          if files != []:
-            print("Could not find the following specified test files:")
-            for f in files:
-              print(f)
-            sys.exit(1)
-          return
-
-        # step 3. form a cut-down list of the xml files matching the correct length and the correct parallelism.
-        working_set = []
-        for xml_file in xml_files:
-          p = etree.parse(xml_file)
-          prob_defn = p.findall("problem_definition")[0]
-          prob_length = prob_defn.attrib["length"]
-          prob_nprocs = int(prob_defn.attrib["nprocs"])
-          if prob_length == length or (length == "any" and prob_length not in ["special", "long"]):
-            if self.parallel == "parallel":
-              if prob_nprocs > 1:
-                working_set.append(xml_file)
-            elif self.parallel == "serial":
-              if prob_nprocs == 1:
-                working_set.append(xml_file)
-            elif self.parallel == "any":
-              working_set.append(xml_file)
-                
-        def get_xml_file_tags(xml_file):
-          p = etree.parse(xml_file)
-          p_tags = p.findall("tags")
-          if len(p_tags) > 0 and not p_tags[0].text is None:
-            xml_tags = p_tags[0].text.split()
-          else:
+
+def add_path_to_environment_variable(env_var, env_path):
+    try:  # Check if the the environment variable already contains the path
+        if str(env_path) not in environ[env_var]:
+            environ[env_var] += f":{env_path}"
+    except KeyError:  # If the environment variable does not exist, create it
+        environ[env_var] = str(env_path)
+
+
+def filter_tests(xml_files):
+    tests = {}
+    # Iterate through all found XML files and keep only the ones that match the
+    # program input arguments
+    for xml_file in xml_files:
+        # Obtain basic information about the test
+        parsed_xml = parse(xml_file).getroot()
+        assert parsed_xml.tag == "testproblem", xml_file
```

Comment:
I was under the assumption that all Fluidity tests, apart from unit tests, are described by an XML file with a `testproblem` root tag. Isn't it the case?

Reply:
Yes, they are but .xml is of course a very common file extension for all sorts of tools (e.g. junit). So it quite likely that users will have some .xml files that are not `testproblem`s somewhere in their test directories even though they're not present in a clean checkout. As far as I can tell this routine is being fed a list of all .xml files (tests/*/*.xml etc.) - so I think when it encounters an xml file that is *not* a `testproblem` it should just silently ignore it rather than fail that assert.